runners/flink/src/test/java/org/apache/beam/runners/flink/streaming/UnboundedSourceWrapperTest.java

```diff
@@ -307,7 +307,10 @@ public void close() {}
         if (seenWatermark.get()) {
           break;
         }
-        Thread.sleep(50);
+
+        // Consider that UnboundedSourceWrapper needs to acquire the checkpoint lock below.
+        // So wait for enough time for that to happen.
+        Thread.sleep(200);
```

Comment:
Can we use a `CountdownLatch` or something like that instead of a sleep?

Reply:
It is tricky because the processing time is expected to be advanced for the event time timers to progress. 

I've refactored the test to wait for the readers to finish; only then the time is advanced and the final trigger executed. That avoids constantly holding the lock.